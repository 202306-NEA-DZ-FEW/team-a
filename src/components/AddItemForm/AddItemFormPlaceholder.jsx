function AddItemFormPlaceholder() {
  return (
    <div className='md:max-w-lg flex flex-col items-center gap-4 mx-auto'>
      <div className='w-[80%] h-[30px] rounded-full bg-gray-300 animate-pulse mb-6' />
      <div className='w-[100%] h-[50px] rounded-full bg-gray-300 animate-pulse' />
      <div className='w-[100%] h-[50px] rounded-full bg-gray-300 animate-pulse' />
      <div className='w-[100%] h-[50px] rounded-full bg-gray-300 animate-pulse' />
      <div className='w-[100%] h-[120px] rounded-md bg-gray-300 animate-pulse' />

      <div className='w-full flex flex-col md:flex-row gap-4'>
        <div className='w-[100%] h-[40px] rounded-md bg-gray-300 animate-pulse' />
        <div className='w-[100%] h-[40px] rounded-md bg-gray-300 animate-pulse' />
        <div className='w-[100%] h-[40px] rounded-md bg-gray-300 animate-pulse' />
      </div>
      <div className='w-full flex flex-col md:flex-row gap-4'>
        <div className='w-[100%] h-[50px] rounded-md bg-gray-300 animate-pulse' />
        <div className='w-[100%] h-[50px] rounded-md bg-gray-300 animate-pulse' />
      </div>
    </div>
  );
}

export default AddItemFormPlaceholder;
