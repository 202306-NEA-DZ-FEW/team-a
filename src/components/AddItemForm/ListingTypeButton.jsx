import React from "react";

function ListingTypeButton({ text, selected, onSelect }) {
  return (
    <button
      onClick={() => onSelect(text)}
      type='button'
      className={`${
        selected === text ? "bg-secondary " : "bg-black"
      } px-3 py-2 flex-1 text-sm font-light rounded-lg text-white`}
    >
      {text}
    </button>
  );
}

export default ListingTypeButton;
