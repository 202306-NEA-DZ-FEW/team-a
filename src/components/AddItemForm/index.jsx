import { collection, doc, serverTimestamp, setDoc } from "firebase/firestore";
import { useFormik } from "formik";
import Image from "next/image";
import Link from "next/link";
import { useRouter } from "next/router";
import Spinner from "public/images/spinner.svg";
import { useState } from "react";
import { toast } from "react-toastify";
import { v4 as uuidv4 } from "uuid";
import * as Yup from "yup";

import useUploadImages from "@/lib/useUploadImages";

import { useAuth } from "@/context/AuthProvider";
import ProtectedLayout from "@/layout/ProtectedLayout";

import ListingTypeButton from "./ListingTypeButton";
import Input from "../Input";
import SelectInput from "../SelectInput";
import TextAreaInput from "../TextAreaInput";
import { db } from "../../lib/firebase";

function AddItemForm({ t, initialLocale, categories, states }) {
  const [images, setImages] = useState([]);
  const [loading, setLoading] = useState(false);
  const { uploadImages } = useUploadImages();
  const { user } = useAuth();
  const router = useRouter();
  const listingTypes = ["exchangeButton", "requestButton", "donationButton"];
  /*  t("addItem:exchangeButton"),
    t("addItem:requestButton"),
    t("addItem:donationButton"), */
  const handleSelect = (selectedType) => {
    formik.setFieldValue("listingType", selectedType);
  };

  const formik = useFormik({
    initialValues: {
      title: "",
      category: "",
      location: "",
      description: "",
      listingType: "",
    },
    validationSchema: Yup.object().shape({
      title: Yup.string().required(t("addItem:titleRequired")),
      category: Yup.string().required(t("addItem:categoryRequired")),
      location: Yup.string().required(t("addItem:locationRequired")),
      description: Yup.string()
        .max(150, "description too long")
        .required(t("addItem:descriptionRequired")),
      listingType: Yup.string().required(t("addItem:itemTagRequired")),
    }),
    onSubmit: async (values, { resetForm }) => {
      const id = uuidv4();
      const imageRef = `items/${id}/${values.title}`;
      if (images.length < 1) {
        toast.error("Min is 1 ❌", {
          position: toast.POSITION.TOP_CENTER,
          autoClose: 1500,
        });
        return;
      }
      if (images.length > 4) {
        toast.error("Max is 4❌", {
          position: toast.POSITION.TOP_CENTER,
          autoClose: 1500,
        });
        return;
      }
      try {
        setLoading(true);
        const colRef = collection(db, "items");
        const imagesUrl = await uploadImages(imageRef, images);
        const valuesWithImages = {
          ...values,
          title: values.title.toLowerCase(),
          id,
          uid: user.uid,
          images: imagesUrl,
          createdAt: serverTimestamp(),
        };
        await setDoc(doc(colRef, id), valuesWithImages);
        toast.success("Item created ✅", {
          position: toast.POSITION.TOP_CENTER,
          autoClose: 1500,
        });
        router.push("/products");
      } catch (err) {
        alert(err.message);
      } finally {
        setLoading(false);
        resetForm();
        setImages([]);
      }
    },
  });
  return (
    <ProtectedLayout>
      <section className='relative md:max-w-lg mx-auto'>
        <form onSubmit={formik.handleSubmit} className='flex flex-col gap-4'>
          <h1 className='text-3xl font-black text-center'>
            {t("addItem:addItem")}
          </h1>
          {loading && (
            <div className='absolute w-full h-full flex items-center justify-center'>
              <Image
                src={Spinner}
                alt='loading'
                height={100}
                width={100}
                className='w-20 h-20'
                priority
              />
            </div>
          )}
          <Input
            name='title'
            type='text'
            placeholder={t("addItem:titlePlaceholder")}
            label={t("addItem:titleLabel")}
            handleChange={formik.handleChange}
            value={formik.values.title}
            handleBlur={formik.handleBlur}
            error={formik.errors.title}
            touched={formik.touched.title}
          />
          <SelectInput
            name='category'
            data={categories}
            label={t("addItem:categoryLabel")}
            handleChange={formik.handleChange}
            value={formik.values.category}
            touched={formik.touched.category}
            error={formik.errors.category}
          />
          <SelectInput
            name='location'
            data={states}
            label={t("addItem:locationLabel")}
            handleChange={formik.handleChange}
            value={formik.values.location}
            touched={formik.touched.location}
            error={formik.errors.location}
          />
          <TextAreaInput
            name='description'
            type='text'
            placeholder={t("addItem:descriptionPlaceHolder")}
            label={t("addItem:descriptionLabel")}
            handleChange={formik.handleChange}
            value={formik.values.description}
            handleBlur={formik.handleBlur}
            error={formik.errors.description}
            touched={formik.touched.description}
            textLength={formik.values.description.length}
            maxLength={150}
          />
          <div className='label'>
            <span htmlFor='file-upload' className='label-text'>
              <span className='text-error'>*</span> {t("addItem:imageLabel")}
            </span>
          </div>
          <input
            dir={initialLocale == "ar" ? "ltr" : ""}
            name='image'
            type='file'
            label={t("addItem:imageLabel")}
            onChange={(e) => {
              setImages(Array.from(e.target.files));
            }}
            className='file-input bg-white'
            id='file-upload'
            multiple
          />
          <div className='form-control'>
            <div className='label'>
              <span className='label-text'>
                <span className='text-error'>*</span> {t("addItem:itemTag")}
              </span>
              <span
                className={`label-text-alt ${
                  formik.touched.listingType && formik.errors.listingType
                    ? "text-error"
                    : ""
                }`}
              >
                {formik.touched.listingType && formik.errors.listingType
                  ? formik.errors.listingType
                  : ""}
              </span>
            </div>
            <div className='flex flex-col md:flex-row gap-2 md:justify-between md:items-center'>
              {listingTypes.map((listingKey, i) => (
                <ListingTypeButton
                  key={i}
                  text={t(`addItem:${listingKey}`)}
                  value={listingKey}
                  selected={formik.values.listingType}
                  onSelect={handleSelect}
                />
              ))}
            </div>
          </div>
          <div className='flex flex-col md:flex-row gap-2 justify-center mt-2'>
            <button
              type='submit'
              className='btn btn-primary flex-1 md:self-center text-white'
            >
              {t("addItem:addItem")}
            </button>
            <Link href='/products' className='btn flex-1 md:self-center'>
              {t("common:buttons:cancel")}
            </Link>
          </div>
        </form>
      </section>
    </ProtectedLayout>
  );
}

export default AddItemForm;
