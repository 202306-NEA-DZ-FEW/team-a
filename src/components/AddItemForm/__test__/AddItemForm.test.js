import renderer from "react-test-renderer";

import AddItemForm from "..";

jest.mock("@/lib/firebase", () => {
  return {
    app: {}, // Mock app object
    db: {}, // Mock db object
    storage: {}, // Mock storage object
    auth: {}, // Mock auth object
  };
});

it("renders correctly", () => {
  const mockCategories = [{ name: "name", stateKey: "stateKey" }];
  const mockStates = [{ name: "name", stateKey: "stateKey" }];
  const tree = renderer
    .create(<AddItemForm categories={mockCategories} states={mockStates} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
