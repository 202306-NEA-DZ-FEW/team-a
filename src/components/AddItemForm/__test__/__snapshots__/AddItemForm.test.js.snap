// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddItemForm snapshot test 1`] = `
<div
  className="min-h-screen"
/>
`;
